lib/rust/ensogl/core/src/display/object/instance.rs

```diff
@@ -1020,139 +1489,1173 @@ impl InstanceDef {
+    }
+}
+
+/// The layout is resolved in two passes, one for the X-axis, one for the Y-axis. See the docs of
+/// this module to learn more.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum LayoutResolutionPass {
+    First,
+    Second,
+}
+
+/// Pass configuration.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum PassConfig {
+    Default,
+    DoNotHugDirectChildren,
+}
+
+trait ResolutionDim = Copy + Debug
+where
+    Vector2<Size>: DimSetter<Self>,
+    Vector2<alignment::OptDim1>: DimSetter<Self>,
+    Vector2<alignment::Dim1>: DimSetter<Self>,
+    Vector2<SideSpacing>: DimSetter<Self>,
+    Vector2<bool>: DimSetter<Self>,
+    Vector2<f32>: DimSetter<Self>,
+    Vector2<Unit>: DimSetter<Self>,
+    Vector3<f32>: DimSetter<Self>,
+    (NonEmptyVec<ColumnOrRow>, NonEmptyVec<ColumnOrRow>):
+        DimRef<Self, Output = NonEmptyVec<ColumnOrRow>>;
+
+
+impl Model {
+    /// Reset size of this display object to values that can be computed statically. This is, use
+    /// the size property if it is set to pixels or to percent of the parent size. Otherwise, set
+    /// the size to zero.
+    fn reset_size_to_static_values<Dim>(&self, x: Dim, parent_size: f32)
+    where Dim: ResolutionDim {
+        let size = match self.layout.size.get_dim(x) {
+            Size::Fixed(unit) => unit.resolve_const_and_percent(parent_size).unwrap_or(0.0),
+            Size::Hug => 0.0,
+        };
+        self.layout.computed_size.set_dim(x, size);
+    }
+
+    fn refresh_layout_internal(&self, pass: LayoutResolutionPass, pass_cfg: PassConfig) {
+        if self.dirty.transformation.check() || self.dirty.modified_children.check_all() {
+            if let Some(layout) = &*self.layout.auto_layout.borrow() && layout.enabled {
+                match pass {
+                    LayoutResolutionPass::First => self.refresh_grid_layout(X, &layout, pass),
+                    LayoutResolutionPass::Second => self.refresh_grid_layout(Y, &layout, pass),
+                }
+            } else {
+                match pass {
+                    LayoutResolutionPass::First => self.refresh_manual_layout(X, pass, pass_cfg),
+                    LayoutResolutionPass::Second => self.refresh_manual_layout(Y, pass, pass_cfg),
+                }
+            }
+        }
+    }
+
+    fn refresh_manual_layout<Dim>(&self, x: Dim, pass: LayoutResolutionPass, pass_cfg: PassConfig)
+    where Dim: ResolutionDim {
+        let hug_children = pass_cfg != PassConfig::DoNotHugDirectChildren;
+        let hug_children = hug_children && self.layout.size.get_dim(x).is_hug();
+        let children = self.children();
+
+        let mut min_x = if children.is_empty() { 0.0 } else { f32::MAX };
+        let mut max_x = if children.is_empty() { 0.0 } else { f32::MIN };
+        let mut children_to_grow = vec![];
+        for child in &children {
+            if child.layout.grow_factor.get_dim(x) > 0.0 {
+                children_to_grow.push(child);
+            } else {
```

Comment:
Wait, are children in manual layout supposed to be growing and filling my space? This is not information I got easily from docs.

Reply:
Added to docs